exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitTest.java

```diff
@@ -31,115 +33,222 @@
+        .build()) {
+      OptionalInt expectedServerPort = OptionalInt.of(serverPort);
+      assertThat(testKit.getServerPort()).isEqualTo(expectedServerPort);
+    }
+  }
+
+  @Test
+  void createTestKitWithDeployedArtifactWithoutCreatedServicesThrows() {
     Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
-    List<Class<? extends ServiceModule>> serviceModules = ImmutableList.of(TestServiceModule.class,
-        TestServiceModule.class);
     TestKit.Builder testKitBuilder = TestKit.builder()
-        .withServices(serviceModules);
-    assertThrows(exceptionType, testKitBuilder::build);
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType, testKitBuilder::build);
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Following service artifacts were deployed, but not used for"
+            + " service instantiation: [%s]", ARTIFACT_ID.toString());
+  }
+
+  @Test
+  void createTestKitWithoutDeployedArtifactThrows() {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID));
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Service %s should be deployed first in order to be created",
+        ARTIFACT_ID.toString());
   }
 
   @Test
-  void createTestKitWithBuilderForMultipleDifferentServices() {
+  void createTestKitWithDifferentDeployedArtifactAndCreatedServiceThrows() {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID_2, SERVICE_NAME_2, SERVICE_ID_2));
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Service %s should be deployed first in order to be created",
+        ARTIFACT_ID_2.toString());
+  }
+
+  @Test
+  void createTestKitWithoutArtifactsDirectoryThrows() {
+    Class<NullPointerException> exceptionType = NullPointerException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID);
+    NullPointerException thrownException = assertThrows(exceptionType, testKitBuilder::build);
+    assertThat(thrownException.getMessage()).isEqualTo("Artifacts directory was not set.");
+  }
+
+  @Test
+  void createTestKitWithBuilderForSingleServiceWithCustomConfiguration() {
```

Comment:
There was a question on configuration failure that results in a service instantiation failure. What would happen in this case?

Reply:
With the latest changes in core the runtime's `.add_service()` implementation should return `Err` and this should fail the `Blockchain` creation (the `BlockchainBuilder#finalize()` -> `dispatcher#add_builtin_service()` -> `dispatcher#add_service()` -> `runtime#add_service()` chain)

Reply:
Thanks, I was a bit confused because I saw that in dispatcher but it was recently moved from there - https://github.com/exonum/exonum/blob/fd3b037c2b71bd1253c0b867687e7510d1e2acb9/exonum/src/runtime/dispatcher/mod.rs#L279